core/src/saros/session/internal/SarosSession.java

```diff
@@ -861,4 +881,62 @@ boolean hasActivityProducers() {
   boolean hasActivityConsumers() {
     return !activeActivityConsumers.isEmpty() || !passiveActivityConsumers.isEmpty();
   }
+
+  /**
+   * Returns whether adding the given reference point to the session would create nested reference
+   * points.
+   *
+   * @param addedReferencePoint the reference point to add to the session
+   * @return whether adding the given reference point to the session would create nested reference
+   *     points
+   */
+  private boolean wouldCreateNestedReferencePoint(IReferencePoint addedReferencePoint) {
+    for (IReferencePoint sharedReferencePoint : getReferencePoints()) {
+      if (referencePointComparator.areNested(addedReferencePoint, sharedReferencePoint)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Logs an error and leaves the session.
+   *
+   * <p>Stops the session using a different thread to prevent lockups.
+   *
+   * <p>This method should be called to end the running session if a user tried to add a new
+   * reference point that would lead to nested reference points being shared.
+   *
+   * @param addedReferencePoint the newly added reference point
+   * @see #wouldCreateNestedReferencePoint(IReferencePoint)
+   */
+  // TODO show user error message
+  private void leaveSessionDueToNestedReferencePoint(IReferencePoint addedReferencePoint) {
```

Comment:
Looking at the negotiation code, throwing just an exception will:

1. stop the session on the remote side (this is what you want)
1.1 this is not true if the receiving side is the host, see https://github.com/saros-project/saros/blob/master/core/src/saros/negotiation/AbstractIncomingResourceNegotiation.java#L303
2. presents an unknown error containing the message of the exception (not the best option but better than this here)

I do not looked what will happen if the sharing side will trigger this, but I think if this happens on the sharing side it will do no harm at all.

I think a possible "fix" is to modify the IPN in such way that it kicks out the remote user when the receiving side is the host.

Reply:
But I don't get what the issue with the current approach is. The session teardown seems to work just fine. I haven't tried non-host project sharing though.

(Also, I would like to avoid putting too much additional work into this as I am moving on to writing the actual thesis part of the thesis.)